File: DbmsGrpcService/Models/TableDifference.cs
namespace DBMS.Models
{
    public class TableDifference : Table
    {
        public readonly Table LeftTable, RightTable;

        private TableDifference(int id, string name, Column[] columns, Table leftTable, Table rightTable)
            : base(id, name, columns)
        {
            LeftTable = leftTable;
            RightTable = rightTable;
        }

        public void Calculate()
        {
            base.ClearRows();
            foreach (Row row in LeftTable.Rows.Values)
                if (!RightTable.ContainsRow(row))
                    base.AddRow(row.Cells);
        }

        public static TableDifference Create(Table leftTable, Table rightTable)
        {
            if (leftTable.Columns.Count != rightTable.Columns.Count)
                throw new ArgumentException("Table difference: tables have different column counts");
            for (int i = 0; i < leftTable.Columns.Count; i++)
                if (leftTable.Columns[i].Type != rightTable.Columns[i].Type)
                    throw new ArgumentException("Table difference: tables have different column types");
            Column[] columns = new Column[leftTable.Columns.Count];
            for (int i = 0; i < leftTable.Columns.Count; i++)
                columns[i] = new(
                    leftTable.Columns[i].Name == rightTable.Columns[i].Name ?
                        leftTable.Columns[i].Name :
                        $"\"{leftTable.Columns[i].Name}\" / \"{rightTable.Columns[i].Name}\"",
                    leftTable.Columns[i].Type);
            TableDifference difference = new(
                leftTable.Id | (rightTable.Id << 16) | (1 << 31),
                $"Difference \"{leftTable.Name}\" - \"{rightTable.Name}\"",
                columns,
                leftTable, rightTable);
            return difference;
        }

        public static TableDifference CreateOrNull(Table leftTable, Table rightTable)
        {
            try
            {
                return Create(leftTable, rightTable);
            }
            catch
            {
                return null;
            }
        }

        public override Row AddRow(Values.Value[] cells) => throw new NotImplementedException("Table difference is read-only");
        public override void RemoveRow(int id) => throw new NotImplementedException("Table difference is read-only");
        public override void ClearRows() => throw new NotImplementedException("Table difference is read-only");

        public override DbmsGrpc.Messages.TableInfo ToMessageInfo()
        {
            var message = base.ToMessageInfo();
            message.TableDifferenceInfo = new()
            {
                LeftTableInfo = LeftTable.ToMessageInfo(),
                RightTableInfo = RightTable.ToMessageInfo()
            };
            return message;
        }

        public override DbmsGrpc.Messages.Table ToMessage()
        {
            var message = base.ToMessage();
            message.TableDifference = new();
            return message;
        }

        public static (int leftId, int rightId)? ParseId(int id)
        {
            if ((id & (1 << 31)) == 0)
                return null;
            id ^= 1 << 31;
            return (id & ((1 << 16) - 1), id >> 16);
        }
    }
}

File: DbmsGrpcPythonServer/Models/TableDifference.py
﻿from Models.Table import Table
from Models.Column import Column
from Protos import Messages_pb2


class TableDifference(Table):
    def __init__(self, id, name, columns, left_table, right_table):
        super().__init__(id, name, columns)
        self.left_table = left_table
        self.right_table = right_table

    def calculate(self):
        super().clear_rows()
        for row in self.left_table.rows.values():
            if not self.right_table.contains_row(row):
                super().add_row(row.cells)

    @staticmethod
    def create(left_table, right_table):
        if len(left_table.columns) != len(right_table.columns):
            raise ValueError("Table difference: tables have different column counts")
        for i in range(len(left_table.columns)):
            if left_table.columns[i].type != right_table.columns[i].type:
                raise ValueError("Table difference: tables have different column types")
        columns = [Column(
            left_table.columns[i].name if left_table.columns[i].name == right_table.columns[i].name
            else f'"{left_table.columns[i].name}" / "{right_table.columns[i].name}"',
            left_table.columns[i].type)
            for i in range(len(left_table.columns))]
        difference = TableDifference(
            left_table.id | (right_table.id << 16) | ~((1 << 31) - 1),
            f'Difference "{left_table.name}" - "{right_table.name}"',
            columns,
            left_table,
            right_table)
        return difference

    @staticmethod
    def create_or_null(left_table, right_table):
        try:
            return TableDifference.create(left_table, right_table)
        except:
            return None

    def add_row(self, cells):
        raise NotImplementedError("Table difference is read-only")

    def remove_row(self, id):
        raise NotImplementedError("Table difference is read-only")

    def clear_rows(self):
        raise NotImplementedError("Table difference is read-only")

    def to_message_info(self):
        return Messages_pb2.TableInfo(Id=self.id, Name=self.name,
                                      TableDifferenceInfo=Messages_pb2.TableDifferenceInfo(
                                          LeftTableInfo=self.left_table.to_message_info(),
                                          RightTableInfo=self.right_table.to_message_info()))

    def to_message(self):
        message = Messages_pb2.Table(Id=self.id, Name=self.name, TableDifference=Messages_pb2.TableDifference())
        message.Columns.extend(column.to_message() for column in self.columns)
        for key, value in self.rows.items():
            message.Rows[key].CopyFrom(value.to_message())
        return message

    @staticmethod
    def parse_id(id):
        if (id & (1 << 31)) == 0:
            return None
        id ^= ~((1 << 31) - 1)
        return id & ((1 << 16) - 1), id >> 16
